src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4277,6 +4277,122 @@ class StubGenerator: public StubCodeGenerator {
     return (address) start;
   }
 
+  void chacha20_quarter_round(VectorRegister aVec, VectorRegister bVec,
+                          VectorRegister cVec, VectorRegister dVec, VectorRegister tmp_vr) {
+    // a += b, d ^= a, d <<<= 16
+    __ vadd_vv(aVec, aVec, bVec);
+    __ vxor_vv(dVec, dVec, aVec);
+    __ vrol_vwi(dVec, 16, tmp_vr);
+
+    // rev32(dVec, T8H, dVec);
+
+    // c += d, b ^= c, b <<<= 12
+    __ vadd_vv(cVec, cVec, dVec);
+    __ vxor_vv(bVec, bVec, cVec);
+    __ vrol_vwi(bVec, 12, tmp_vr);
+
+    // a += b, d ^= a, d <<<= 8
+    __ vadd_vv(aVec, aVec, bVec);
+    __ vxor_vv(dVec, dVec, aVec);
+    __ vrol_vwi(dVec, 8, tmp_vr);
+
+    // c += d, b ^= c, b <<<= 7
+    __ vadd_vv(cVec, cVec, dVec);
+    __ vxor_vv(bVec, bVec, cVec);
+    __ vrol_vwi(bVec, 7, tmp_vr);
+  }
+
+  /**
+   * int com.sun.crypto.provider.ChaCha20Cipher.implChaCha20Block(int[] initState, byte[] result)
+   *
+   *  Input arguments:
+   *  c_rarg0   - state, the starting state
+   *  c_rarg1   - key_stream, the array that will hold the result of the ChaCha20 block function
+   */
+  address generate_chacha20Block() {
+    Label L_Rounds;
+
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20Block");
+    address start = __ pc();
+
+    const int states_len = 16;
+    const int step = 4;
+    const Register state = c_rarg0;
+    const Register key_stream = c_rarg1;
+    const Register loop = t0;
+    const Register tmp_addr = t1;
+    const Register length = t2;
+    const Register avl = x28;
+    const Register stride = x29;
+
+    const VectorRegister work_vrs[16] = {
+      v4,  v5,  v6,  v7,  v16, v17, v18, v19,
+      v20, v21, v22, v23, v24, v25, v26, v27
+    };
+    const VectorRegister tmp_vr = v29;
+    const VectorRegister counter = v30;
+
+
+    RegSet saved_regs;
+    __ push_reg(saved_regs, sp);
+
+    // Put 16 here, as com.sun.crypto.providerChaCha20Cipher.KS_MAX_LEN is 1024
+    // in java level.
+    __ li(avl, 16);
+    __ vsetvli(length, avl, Assembler::e32, Assembler::m1, Assembler::ma, Assembler::ta);
```

Comment:
Is this really correct.
We have no uses of ma/ta before this since we need to make sure we never touch memory outside of the arrays.
I don't think ma/ta will ever be correct when working on Java heap.

I would drop the last two argument and use the default of mu/tu as we do everywhere else.

Reply:
I'm not quite sure, but modified as you suggested.

Reply:
There is some statement at https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#343-vector-tail-agnostic-and-vector-mask-agnostic-vta-and-vma, 
```
The agnostic policy was added to accommodate machines with vector register renaming. With an undisturbed policy, all elements would have to be read from the old physical destination vector register to be copied into the new physical destination vector register. This causes an inefficiency when these inactive or tail values are not required for subsequent calculations.
```
Seems it's more effiecient at some situation, but I'm not sure what's that case mentioned above.

Reply:
At https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#7-vector-loads-and-stores, there is a statement, ```Vector loads and stores can be masked, and they only access memory or raise exceptions for active elements.```
Seems it will not "touch memory outside of the arrays" when store back to key stream array(which in java heap)

Reply:
Anyway, I think your suggestion make more sense, I've modified as you suggested.
Thanks Robbin for the detailed reviewing!

Reply:
It's the paragraph above:
`The tail elements during a vector instruction’s execution are the elements past the current vector length setting specified in vl.`

This means they are outside of your working set (body).

`When a set is marked agnostic, the corresponding set of destination elements in any vector destination operand can either retain the value they previously held, or are overwritten with 1s. Within a single vector instruction, each destination element can be either left undisturbed or overwritten with 1s, in any combination, and the pattern of undisturbed or overwritten with 1s is not required to be deterministic when the instruction is executed with the same inputs.`

Maybe this can't happen here since you work with such nice size.
But the arrays result and startState will be initialized to zero when allocated.
We don't want anything changing bytes which the VM didn't write to and we must be sure never to write outside of these. 

So tail is what comes after you data, in our case that will be Java heap, so we must always be sure only to write to the body.

Did that make sense?

Reply:
Thanks for discussion. Yes, I see your point.
What I mean above is that the java heap can only be touched when store back to memory, but store `only access memory or raise exceptions for active elements.`, that seems mean that a vectore operation might touch tail elements in vector group register, but a vector store will not touch the memory corresponding to the tail elements.
Am I understanding the spec correctly?

Reply:
I don't see that the specs guarantees that under no circumstance tail memory will not be touched.
So I don't know, for now it's better to be safe than sorry.

Reply:
> Maybe this can't happen here since you work with such nice size.

Yeah, but given we don't know what the vector register size is I don't think we can assume anything about how "nice" the size is. ;-) I agree with Robbin here, I think mu / tu is the right approach.

Reply:
> I don't think ma/ta will ever be correct when working on Java heap.

What if we know that the operation is definitely touching memory that's inside the object? And why is it different to working on, say, the C-heap? You can have other objects on either side in both cases.